Process disclosures properly (#245)
oauth-wg/oauth-selective-disclosure-jwt: draft-ietf-oauth-selective-disclosure-jwt.md

    2. Validate the signature over the SD-JWT.
    3. Validate the Issuer of the SD-JWT and that the signing key belongs to this Issuer.
    4. Check that the SD-JWT is valid using `nbf`, `iat`, and `exp` claims, if provided in the SD-JWT, and not selectively disclosed.
    5. Check that the `_sd_alg` claim value is understood and the hash algorithm is deemed secure.
 4. Process the Disclosures and `_sd` keys in the SD-JWT as follows:

>     1. Create a copy of the SD-JWT payload, if required for further processing.

Reviewer: I think I asked before but what is the purpose here?

Author: This might be required depending on how the processing is implemented. It's an implementation detail, but hopefully helps with understanding the algorithm.

Reviewer: #247

>     2. For each Disclosure provided:
>        1. Calculate the digest over the base64url string as described in (#hashing_disclosures).

Reviewer: ```suggestion
    2. For each Disclosure provided:
        1. If the Disclosure is not a JSON-encoded array of three elements, the Verifier MUST reject the Presentation.
        2. Calculate the digest over the base64url-encoded string as described in (#hashing_disclosures).
```
the rest of the document seem to use `base64url-encoded`, we might want to do a clean up PR on that if not in this one

Author: base64url-encoded works for me.

Reviewer 2: base64url-encoded works for me too. So it'd be
```suggestion
    2. For each Disclosure provided:
       1. Calculate the digest over the base64url-encoded string as described in (#hashing_disclosures).
```

>     3. Find all `_sd` keys in the SD-JWT payload. For each such key:

Author: Add marker so that we can refer to this later:

```suggestion
    3. Find all `_sd` keys in the SD-JWT payload. For each such key perform the following steps (*):
```

       1. If the key does not refer to an array, the Verifier MUST reject the Presentation.
       2. Otherwise, process each entry in the `_sd` array as follows:

>           1. Using the previously calculated digests, find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.

Reviewer: ```suggestion
          1. Compare the value with the digest calculated in an earlier processing step, and find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.
```

Reviewer 2: I think digest would be plural in that suggested change? i.e for the value from the _sd array, compare it to each of the calculated Disclosure digests to find the match. 

"Compare the value with the digests calculated in an earlier processing step..."

Reviewer 2: ```suggestion
          1. Compare the value with the digests calculated in an earlier processing step, and find the matching Disclosure. If no such Disclosure can be found, the digest MUST be ignored.
```

Author: In *an* earler processing step is a bit unclear. How about just "previously"?

>           2. If the Disclosure is not a JSON-encoded array of three elements, the Verifier MUST reject the Presentation.

Reviewer: shouldn't this belong to a `2. For each Disclosure provided:` section?
```suggestion
```

Reviewer 2: I think it could go in either spot but having it here might be better from a security/integrity perspective because it suggests not to do anything with the content of a Disclosure until/unless it's digest value is found in an _sd element.

Author: My reasoning for putting that check here was that until here, there is no need to JSON-parse the disclosures. It feels more natural to me to check before further processing.

Reviewer: > it suggests not to do anything with the content of a Disclosure until/unless it's digest value is found in an _sd element.

ok, agreed to keep this sentence here.

          3. Insert, at the level of the `_sd` key, a new claim using the claim name and claim value from the Disclosure.
          4. If the claim name already exists at the same level, the Verifier MUST reject the Presentation.

>           5. If the decoded value contains an `_sd` key in an object, recursively process the key as described here.

Reviewer: ```suggestion
          5. If the decoded value contains an `_sd` key in an object, recursively process each entry in the `_sd` array by repeating these steps 1 to 5.
```
`described here` can mean a lot of things...

    3. Validate the Issuer of the SD-JWT and that the signing key belongs to this Issuer.
    4. Check that the SD-JWT is valid using `nbf`, `iat`, and `exp` claims, if provided in the SD-JWT, and not selectively disclosed.

>     5. Check that the `_sd_alg` claim is present and its value is understood and the hash algorithm is deemed secure.

Reviewer 2: ```suggestion
    5. Check that the `_sd_alg` claim value is understood and the hash algorithm is deemed secure.
```

it's optional with a default now so removing the "is present" part

Pieter Kasselman,
Ryosuke Abe,
Shawn Butterfield,
Filip Skokan,

> Christian Bormann,

Author: ```suggestion
Christian Bormann,
```

> Christian Bormann,

Reviewer: @bc-pi can you please put in the alphabetic order? `Matthew Miller` seems to be out of place too.

Reviewer 2: Will do. Sorry, I somehow missed that they were in alphabetic order on first name (guess I was looking at last).

Reviewer 2: done w/ ca20ac44bdc0c2d9022256ad7526464d67c73620
